### ISSUE TITLE
Application.onCreate lint check warns when MParticle is initialized in the Application through a series of nested function calls

### ISSUE THREAD

**Issue by erawhctim** [mParticle/mparticle-android-sdk#45]
The lint check to ensure that `MParticle.start` is called in `Application.onCreate` incorrectly warns on our project, due to the fact that we use a wrapper class in our Application. The call chain roughly looks like this:

```
class MyApplication() { 
    fun onCreate() {
        ...
        Analytics.initialize()
    }
}
```

```
class Analytics() { 
    fun initialize() {
        ...
        initMParticle()
    }

    private fun initMParticle() {
        ...
        // Lint warning here
        MParticle.start(
            {Builder init}
        )
    }
}
```

Is there any way to modify the lint check to walk up the tree of function calls in the AST and figure out if `MParticle.start` is _eventually_ called in the Application class?

**Comment by willpassidomo** [comment]
It does walk up the tree, although we have an arbitrary cut off at 5 levels in just to make sure we don't get stuck in a recursive walk. It looks like your start function is only 3 levels into the call stack, so we may be inadvertently double counting on our way down. I'm going to test this scenario and get back to you

**Comment by willpassidomo** [comment]
@erawhctim I tested using about the same code as the sample you provided, and the linter picked up that the `start()` call is rooted in the Application (no warnings). I know there might be more to your actual code than the sample, so if you find you are actually more than 5 steps up the call stack and that is the issue, we could raise the limit to 8 or 10 and see how it goes. If that is not the case:
1) does your `Application` subclasses' `onCreate()` method also have mParticle lint warning? It would say "The Method should call MParticle.start()"
2) if not, do you have multiple places (in an active code path or not) that call `MParticle.start()`?

**Comment by willpassidomo** [comment]
any update on this @erawhctim ?

**Comment by erawhctim** [comment]
@willpassidomo Sorry for the lack of follow-up here, and thanks for digging into this issue. 

Our `onCreate` does also have the lint warning; it shows up in our application class and within the function that wraps MParticle initialization. 

The precise call chain is this: `Application.onCreate() { Analytics.initialize }`, `Analytics.initialize() { initMParticle() }`, `initMParticle() { MParticle.start() }`. I see the lint warning on the private `Analytics.initMParticle()`, even with `@SuppressLint("MParticleInitialization")` (which seems wrong). 

Not sure what I'm missing.

**Comment by BasheerMalaaNBCU** [comment]
I see the same issue, why this issue is closed when it's not resolved ?

**Comment by samdozor** [comment]
@BasheerMalaaNBCU looks like we closed this due to lack of response if you look above. But then the response came in right after :-) I'll reopen.

**Comment by BasheerMalaaNBCU** [comment]
@samdozor thanks a lot. Let me know if you need any info. :)

**Comment by samdozor** [comment]
Actually - if you have an example stack trace that seems to reproduce the issue (you can sanitize/change your class names) that would be great.

**Comment by BasheerMalaaNBCU** [comment]
@samdozor The issue exactly the same to what @willpassidomo described. 
From `Application.onCreate()` we call `AnalyticsTracker.initialize()`
Inside `AnalyticsTracker.initialize()` we initialize `MParticleWrapper()` and inside  `MParticleWrapper()` we call `Mparticle.start(..Options..)`

**Comment by willpassidomo** [comment]
Hey @BasheerMalaaNBCU @erawhctim sorry for the late update, but this issue should be resolved in version `5.14.16`. I'm closing, for now, let me know if you still see the issue

**Comment by hikaritenchi** [comment]
I wasn't sure whether to create a new issue or revive this one, but we're having a similar issue in `5.21.0`, although I will admit that our whole `onCreate()` chain does look pretty busy...

```kotlin
open class MyApplication : Application() {
  //This Method should call MParticle.start()
  override fun onCreate() {
    ApplicationModule.analyticsTracker.initialize()
  }
}
```
```kotlin
object ApplicationModule {
  val analyticsTracker by lazy { AnalyticsTracker() }
  fun initialize() { }
}
```
```kotlin
class AnalyticsTracker() {
  private val mParticle = MParticleWrapper()
}
```
```kotlin
class MParticleWrapper {
  init {
    start()
  }
  private fun start() {
    //MParticle.start() should be called in Application.onCreate(), not here
    MParticle.start(options.build())
  }
}
```
